projects/bootloader/src/ping.c

```diff
@@ -0,0 +1,44 @@
+#include "ping.h"
+
+#include "bootloader_can.h"
+#include "bootloader_datagram_defs.h"
+#include "can_datagram.h"
+#include "config.h"
+#include "dispatcher.h"
+
+// this function needs to be declared for response_config datagram
+static StatusCode prv_tx_callback(uint8_t *data, size_t len, bool is_start_message);
+// response_config datagram setup
+static uint8_t board_id;
+static uint8_t destination_node = 0;
+static CanDatagramTxConfig response_config = {
+  .dgram_type = BOOTLOADER_DATAGRAM_PING_RESPONSE,
+  .destination_nodes_len = 1,
+  .destination_nodes = &destination_node,
+  .data_len = 1,
+  .data = &board_id,
+  .tx_cb = prv_tx_callback,
+};
```

Comment:
There's a little bit of a gotcha here. As I understand it (@mitchellostler correct me if I'm wrong), after `can_datagram` has TXed everything from the ping response message (which could take several iterations of the event loop), we need to call `can_datagram_start_listener` again so that we start listening for datagrams again.

A good place to do that would be in the TX completion callback. Since every bootloader operation will need to do this, I think the best thing to do would be to add a function to `dispatcher` which you can pass to `tx_cmpl_cb` here which just calls `can_datagram_start_listener`. Then when a module wants to send a datagram, it just has to pass that function as the `tx_cmpl_cb` and doesn't have to worry about calling `can_datagram_start_listener` itself.

Reply:
Yeah, both can_datagram_start_listener disables after completion of one message, then calls the callback. The implementation in the new patch set looks good. The only thing to consider is that if a TX is started then nothing can be rxed, but this looks like it shouldn't be an issue with the current changes